### PR TITLE
Build otherlibs (and other installed artifacts) using our stdlib

### PR DESCRIPTION
**Pull request by dkalinichenko-js**
Split the last stage of our compiler build into two: `compiler` and `main`. `compiler` mimics `boot-compiler`, but uses our standard library instead of the upstream version to build the targets. `main` builds the rest of the installation using that compiler, reusing artifacts already built.

This is necessary to compile `Stdlib_alpha.Or_null` in bytecode correctly: the upstream standard library can't marshal null constants. This is also helpful for supporting `float32` in bytecode.

Note that `unix` is still compiled using `boot-compiler`, since `compiler` depends on `unix`.

It's complicated to add a test for this, since `or_null` in bytecode is not implemented yet, but I successfully built my WIP bytecode branch with this change. I also inspected `_build/log` to ensure it uses the right compiler.

cc @TheNumbat

### CODE REVIEW

**Comment by mshinwell** [comment]
I'll look at this - there have been some problems with 5.2 which might be related.

**Comment by dkalinichenko-js** [comment]
I'm looking into the CI failures, can't reproduce them on my machine so far.

**Comment by mshinwell** [comment]
@dkalinichenko-js can you please get CI to pass on this PR?

**Comment by dkalinichenko-js** [comment]
Seems like Dune clears the `_build/main` default directory between builds in the CI for some reason, while it doesn't locally. I fixed this by hardlinking the resulting compiler instead of symbolic linking. This is not ideal, since we want Dune to reuse already built artifacts in `_build/main`, but should be sufficient for testing.

**Comment by mshinwell** [comment]
Can we try to do something to avoid hardlinking tricks, which can be fragile and difficult to understand?

**Comment by mshinwell** [comment]
(I think the change in behaviour depends on the Dune version, we've encountered this elsewhere iirc)

**Comment by dkalinichenko-js** [comment]
I'll try that, agree it's fragile and bad.

**Comment by dkalinichenko-js** [comment]
I updated the PR to build actual final targets instead of boot targets, because boot targets don't depend on our version of unix and it's very tricky to make them to. I also removed hardlinking in favor of copying. I'm not quite happy with this, since we perform duplicate work if Dune clears `_build/main` between builds, but if the impact on build times is small this should unblock various bytecode changes.

File: Makefile.common-jst [hunk 1]
```diff
@@ -76,9 +89,29 @@ runtime-stdlib: boot-compiler
 # to a separate directory.
 	$(cpl) -R _build/install/runtime_stdlib _build/runtime_stdlib_install
 # Dune does not believe the compiler can make .cmxs unless the following file exists.
-	@touch _build/runtime_stdlib_install/lib/ocaml_runtime_stdlib/dynlink.cmxa
+# Create [dynlink.cmxa] for 4.14 and [dynlink/dynlink.cmxa] for 5.0
+	mkdir -p _build/runtime_stdlib_install/lib/ocaml_runtime_stdlib/dynlink
+	@touch _build/runtime_stdlib_install/lib/ocaml_runtime_stdlib/{,dynlink/}dynlink.cmxa
```
Comment (mshinwell) [inline Makefile.common-jst:94]: Should this change be in a separate PR?

File: Makefile.common-jst [hunk 2]
```diff
@@ -76,9 +89,29 @@ runtime-stdlib: boot-compiler
...
+compiler_targets = \
+  $(ocamlc) \
+  $(ocamlopt) \
+  $(boot_ocamlmklib) \
+  $(boot_ocamldep) \
+  $(boot_ocamlobjinfo)
 
 compiler: runtime-stdlib
+# Build the same target as in [boot-compiler], but with our runtime stdlib.
+# Uses the [main] context to avoid duplicating work.
```
Comment (mshinwell) [inline Makefile.common-jst:105]: I think this "Uses the..." comment is in the wrong place.

File: Makefile.common-jst [hunk 3]
```diff
@@ -76,9 +89,29 @@ runtime-stdlib: boot-compiler
...
+	RUNTIME_DIR=$(RUNTIME_DIR) $(dune) build $(ws_compiler) $(compiler_targets)
+
+main: _build/compiler_install compiler
+# Build the rest of the installation using the compiler from the [main] context.
+# The [boot-compiler] is not sufficient, since it marshals bytecode constants
+# incorrectly, using the upstream stdlib. This miscompiles code using new constants
+# like null pointers or float32 values.
+# However, otherwise [boot-compiler] produces identical artifacts to [compiler],
+# and it's safe-ish to reuse the [main] context again.
```
Comment (mshinwell) [inline Makefile.common-jst:114]: We need a better argument than "safe-ish".  Which artifacts are required from the context in which the final compiler is built in order to build the otherlibs (minus `unix`)?  I'm wondering if we can use a separate context for this last part.

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: Makefile.common-jst (hunk 1 above)
Comment (dkalinichenko-js) [reply Makefile.common-jst:94]: Previously, Dune assumed the 4.14 directory structure for the `boot-compiler`. Now, it also assumes the 5.0 directory structure for the `compiler` we use to build otherlibs, so we need this change to build everything.

File: Makefile.common-jst (hunk 3 above)
Comment (dkalinichenko-js) [reply Makefile.common-jst:113]: Moved the entire `dune_compiler_context` to its own directory. This means we don't reuse any artifacts, potentially slowing down build times but ensuring we do not miscompile anything. I believe, since `main` was already cleared between builds in the CI, this change should not introduce any extra delay there.
